src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1493,19 +1493,22 @@
 
 ;; Wrap the fact-type so that Clojure equality and hashcode semantics are used
 ;; even though this is placed in a Java map.
-(deftype AlphaRootsWrapper [roots roots-hash]
+(deftype AlphaRootsWrapper [fact-type fact-type-hash roots]
   Object
   (equals [this other]
     ;; Since we memoize the function that retrieves these on the fact type, and since the alpha roots for
     ;; any given type primary type T (excluding descendants and ancestors) are a disjoint
     ;; set with respect to the alpha roots for any other primary type T2, we know that we will
     ;; only have one AlphaRootsWrapper object for any given set of alpha roots per get-alphas-fn instance.
-    ;; Therefore we can just use reference equality in the interest of maximizing performance.
-    (identical? this other))
+    ;; Therefore we can reference equality in the interest of maximizing performance and fallback to equivalence,
+    ;; as fact-types can be defined by consumers and thus might not be identical.
+    (let [other ^AlphaRootsWrapper other]
+      (and (= fact-type-hash (.fact-type-hash other))
```

Comment:
Using `=` is adding a lot of overhead to this comparison.  What is the type here?  I think it is `Long`.  I'd suggest to use `==` if you are sure neither are nil, otherwise do so nil-safely.